rfcs/2021-02-23-6531-performance-testing.md

```diff
@@ -499,3 +504,203 @@ pursue [Run Vector Continuously](#run-vector-continuously) in the future but the
+
+#### Reduce Noise
+
+In order to drive stability in our results we have to reduce noise. This is done
+partially by careful configuration of our VMs and elimination of unreliable VMs
+before benchmarks begin -- see `diagnostic` work below -- but because some tests
+will be subject to nondeterminism outside our control we'll need to adjust for
+this on a test by test basis. We expect, especially for the first test, to
+primarily adjust by the use of repeat and parallel runs. After the fact analysis
+of runs can be done to summarize the result, in much the same way as our
+criterion benchmarks function.
+
+Completion criteria: We will be able to run our performance tests in a way that
+reduce noise automatically, based on detected instability. Criterion is our
+model. User's configuration knobs will set details surrounding max parallelism,
+total runtime and etc.
+
+#### Backfill Performance Tests
+
+The value of our performance tests comes in their comparison with previous
+results. We must have the ability to "backfill" results for new benchmarks or
+for when we substantially change our benchmark approach.
+
+Completion criteria: We will backfill our performance tests far back enough to
+give a reasonable trend-line. In cases where compatibility has been broken
+between vector releases we'll support alternate configuration where reasonable.
+
+#### Expand Side-Channel Data
+
+To this point the only data we've collected with any serious intention have been
+the **primary sources of measure**. These measurements give us a notion of how
+the project is trending but does not necessarily give us any guidance as to why
+or what to do about it. We will collect
+[perf](https://perf.wiki.kernel.org/index.php/Main_Page) data from our test
+runs, with the key insights being dependent on the test. We will collect
+flamegraphs of our runs. Likewise, we will collect Vector's internal telemetry
+into a comparable state. We may also explore eBPF tracing to track syscall
+counts.
+
+Completion criteria: Our aim with collecting side-channel data is to provide
+comparison between versions. If, for instance, the test is sensitive to context
+switches how have context switches changed between versions and to what degree
+can we inject new telemetry into Vector to measure (and reduce) this behavior?
+While collecting information is useful the main result here will be tooling to
+_compare_ versions, to demonstrate change.
+
+#### Expand Test Coverage
+
+Once we have pursued a single test to its utmost we must expand out. The
+existing pool of performance tests in test-harness are suitable for conversion,
+though we might consider a new variant if user needs are evident.
+
+Completion criteria: This work will be "complete" when we have five total tests
+in the new test-harness style. Realistically we will continually add onto the
+harness until such time as it no longer proves useful to us, but we gotta draw
+the line somewhere.
+
+### Pursue `diagnostic` Command
```

Comment:
cc @lukesteensen based on our conversations earlier today.

Reply:
The open question here is mostly about workflow. A simple subcommand doesn't provide an obvious way to reach into an existing long-running process, and as much as possible it'd be nice to align this diagnostic data with our existing internal metrics system (i.e. we want that system to be useful for this purpose). The goals remain the same, but we should be open-minded about the actual implementation strategy.